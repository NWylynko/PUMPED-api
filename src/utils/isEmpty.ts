export const isArrayEmpty = (array: any[]): true | undefined => (array.length === 0 ? true : undefined);

export const isEmpty = (object: any): true | undefined => (!object ? true : undefined);
