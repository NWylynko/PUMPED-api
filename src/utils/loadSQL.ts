import fs from 'fs';
import path from 'path';

const loadSQL = (file: string) => fs.readFileSync(path.resolve(__dirname, '../../SQL/', `${file}.SQL`), 'utf-8');

export default loadSQL;
