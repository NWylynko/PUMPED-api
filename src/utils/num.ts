const num = (n: string) => parseInt(n, 10);

export default num;
