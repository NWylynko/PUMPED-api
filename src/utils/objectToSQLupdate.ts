import { escape } from 'sqlstring';

const objectToSQLupdate = (fields: any) => {
  let sql = '';

  Object.keys(fields).forEach((partName, index, array) => {
    // if the current item is the length of the array then its the last item
    // the last item cant have a comma
    const comma = index === array.length - 1 ? '' : ',';

    sql += `${escape(partName)} = ${escape(fields[partName])}${comma} `;
  });

  return sql;
};

export default objectToSQLupdate;
