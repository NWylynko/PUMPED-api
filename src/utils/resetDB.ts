import { promises as fs } from 'fs';
import db, { SQL } from '../db';

export const createTables = () => Promise.all([
  db.run(SQL.tables.Brand),
  db.run(SQL.tables.Collection),
  db.run(SQL.tables.Colour),
  db.run(SQL.tables.ColourImage),
  db.run(SQL.tables.Customer),
  db.run(SQL.tables.Image),
  db.run(SQL.tables.Order),
  db.run(SQL.tables.OrderItem),
  db.run(SQL.tables.Review),
  db.run(SQL.tables.Section),
  db.run(SQL.tables.Shoe),
  db.run(SQL.tables.ShoeTag),
  db.run(SQL.tables.Stock),
  db.run(SQL.tables.Style),
  db.run(SQL.tables.Tag),
  db.run(SQL.tables.WishList),
]);

const resetDB = async () => {
  await db.close();
  await fs.unlink(`./main.${process.env.NODE_ENV}.db`);
  await db.open();
  await createTables();
};

export const addTestData = () => db.exec(SQL.testData);

export const resetDBForTest = async () => {
  await db.close();
  await db.open();
  await createTables();
  await db.exec(SQL.testData);
};

export default resetDB;
