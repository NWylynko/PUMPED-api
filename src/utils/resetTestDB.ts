import { promises as fs } from 'fs';
import db from '../db';

const reset = async () => {
  await db.close();
  await fs.unlink('./main.test.db');
  await fs.copyFile('./main.testData.db', './main.test.db');
  await db.open();
};

export default reset;
