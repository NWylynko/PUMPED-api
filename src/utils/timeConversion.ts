function timeConversion(milliseconds: number) {
  // adapted from https://stackoverflow.com/a/32180863

  const seconds = (milliseconds / 1000);
  const minutes = (milliseconds / (1000 * 60));
  const hours = (milliseconds / (1000 * 60 * 60));
  const days = (milliseconds / (1000 * 60 * 60 * 24));

  if (seconds < 60) {
    return `${Math.floor(seconds)} Seconds`;
  } if (minutes < 60) {
    return `${Math.floor(minutes)} Minutes`;
  } if (hours < 24) {
    return `${Math.floor(hours)} Hours`;
  }
  return `${Math.floor(days)} Days`;
}

export default timeConversion;
