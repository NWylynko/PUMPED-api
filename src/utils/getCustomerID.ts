import { Request } from 'express';
import num from './num';

const getCustomerID = (req: Request): number => {
  const CustomerID = req.header('CustomerID');

  if (!CustomerID) {
    throw new Error('no customer id supplied');
  }

  return num(CustomerID);
};

export default getCustomerID;
