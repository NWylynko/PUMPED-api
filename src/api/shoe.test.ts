// eslint-disable-next-line import/no-extraneous-dependencies
import supertest from 'supertest';
import express from '../app';
import { resetDBForTest } from '../utils/resetDB';

const app = supertest(express);

beforeEach(async (done) => { await resetDBForTest(); done(); });

describe('GET all shoes', () => {
  it('gets all shoes', async () => {
    const response = await app.get('/shoe');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });

  it('with brand of nike', async () => {
    const response = await app
      .get('/shoe?brand=nike');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });

  it('with style of basketball', async () => {
    const response = await app
      .get('/shoe?style=basketball');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });

  it('with brand of nike and style of basketball', async () => {
    const response = await app
      .get('/shoe?brand=nike&style=basketball');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });

  it('with section of mens', async () => {
    const response = await app
      .get('/shoe?section=mens');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });

  it('with collection of hypes', async () => {
    const response = await app
      .get('/shoe?collection=hypes');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });

  it('with section of mens and collection of hypes', async () => {
    const response = await app
      .get('/shoe?section=mens&collection=hypes');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });

  it('with stars over 3', async () => {
    const response = await app
      .get('/shoe?stars=3');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });

  it('with brand nike and stars over 3', async () => {
    const response = await app
      .get('/shoe?brand=nike&stars=3');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });

  it('with non existent brand', async () => {
    const response = await app
      .get('/shoe?brand=notAShoeBrand');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();
  });
});

describe('GET single shoe', () => {
  [1, 2, 3, 4].forEach((id) => {
    it(`get shoe with id ${id}`, async () => {
      const response = await app
        .get(`/shoe/${id}`);

      expect(response.status).toMatchSnapshot();

      const json = JSON.parse(response.text);

      expect(json).toMatchSnapshot();
    });
  });
});

describe('Post add shoe', () => {
  it('no data', async () => {
    const response = await app
      .post('/shoe');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect({ error: json.error, message: json.message }).toMatchSnapshot();
  });

  it('perfect amount of data', async () => {
    const shoe = {
      name: 'test shoe',
      description: 'runs fast',
      price: 50,
      releaseDate: 29456887,
      BrandID: 2,
      StyleID: 1,
      SectionID: 1,
      CollectionID: 1,
      CoverImage: 1,
    };

    const addData = await app
      .post('/shoe')
      .type('json')
      .send(JSON.stringify(shoe));

    expect(addData.status).toMatchSnapshot();

    const addDataJson = JSON.parse(addData.text);

    expect(addDataJson).toMatchSnapshot();

    const getShoe = await app
      .get(encodeURI(`/shoe?name=${shoe.name}`));

    expect(getShoe.status).toMatchSnapshot();

    const getShoeJson = JSON.parse(getShoe.text);

    expect(getShoeJson).toMatchSnapshot();
  });

  it('no cover image', async () => {
    const shoe = {
      name: 'test shoe',
      description: 'runs fast',
      price: 50,
      releaseDate: 29456887,
      BrandID: 2,
      StyleID: 1,
      SectionID: 1,
      CollectionID: 1,
    };

    const addData = await app
      .post('/shoe')
      .type('json')
      .send(JSON.stringify(shoe));

    expect(addData.status).toMatchSnapshot();

    const addDataJson = JSON.parse(addData.text);

    expect(addDataJson).toMatchSnapshot();

    const getShoe = await app
      .get(encodeURI(`/shoe?name=${shoe.name}`));

    expect(getShoe.status).toMatchSnapshot();

    const getShoeJson = JSON.parse(getShoe.text);

    expect(getShoeJson).toMatchSnapshot();
  });
});

describe('PATCH (update) a shoe', () => {
  it('no data', async () => {
    const response = await app
      .patch('/shoe/1');

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect({ error: json.error, message: json.message }).toMatchSnapshot();
  });

  it('1 item updated', async () => {
    const shoe = {
      name: 'fly kicks',
    };

    const ID = 1;

    const updateData = await app
      .patch(`/shoe/${ID}`)
      .type('json')
      .send(JSON.stringify(shoe));

    expect(updateData.status).toMatchSnapshot();

    const updateDataJson = JSON.parse(updateData.text);

    expect(updateDataJson).toMatchSnapshot();

    const getShoe = await app
      .get(encodeURI(`/shoe/${ID}`));

    expect(getShoe.status).toMatchSnapshot();

    const getShoeJson = JSON.parse(getShoe.text);

    expect(getShoeJson).toMatchSnapshot();
  });

  it('3 items updated', async () => {
    const shoe = {
      name: 'test shoe',
      description: 'runs fast',
      price: 50,
    };

    const ID = 3;

    const updateData = await app
      .patch(`/shoe/${ID}`)
      .type('json')
      .send(JSON.stringify(shoe));

    expect(updateData.status).toMatchSnapshot();

    const updateDataJson = JSON.parse(updateData.text);

    expect(updateDataJson).toMatchSnapshot();

    const getShoe = await app
      .get(encodeURI(`/shoe/${ID}`));

    expect(getShoe.status).toMatchSnapshot();

    const getShoeJson = JSON.parse(getShoe.text);

    expect(getShoeJson).toMatchSnapshot();
  });
});

describe('DELETE a shoe', () => {
  it('remove a shoe', async () => {
    const ID = 1;

    const response = await app
      .delete(encodeURI(`/shoe/${ID}`));

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();

    const getShoe = await app
      .get(encodeURI(`/shoe/${ID}`));

    expect(getShoe.status).toMatchSnapshot();

    const getShoeJson = JSON.parse(getShoe.text);

    expect(getShoeJson).toMatchSnapshot();
  });

  it('remove a non existent shoe', async () => {
    const ID = 4;

    const response = await app
      .delete(`/shoe/${ID}`);

    expect(response.status).toMatchSnapshot();

    const json = JSON.parse(response.text);

    expect(json).toMatchSnapshot();

    const getShoe = await app
      .get(encodeURI(`/shoe/${ID}`));

    expect(getShoe.status).toMatchSnapshot();

    const getShoeJson = JSON.parse(getShoe.text);

    expect(getShoeJson).toMatchSnapshot();
  });
});
