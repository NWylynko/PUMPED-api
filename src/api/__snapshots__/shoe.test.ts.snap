// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET all shoes gets all shoes 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "red",
          "hex": "ff0000",
        },
        Object {
          "colour": "blue",
          "hex": "0000ff",
        },
      ],
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 2,
      "Section": "mens",
      "Stars": 2,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "green",
          "hex": "00ff00",
        },
      ],
      "description": "dive",
      "name": "jumpers",
      "price": 80,
      "releaseDate": 2098475298,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 3,
      "Section": "womens",
      "Stars": 0,
      "Style": "runners",
      "colours": Array [],
      "description": "sprint",
      "name": "kicks",
      "price": 120,
      "releaseDate": 2098475234,
    },
  ],
  "success": true,
}
`;

exports[`GET all shoes with brand nike and stars over 3 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "red",
          "hex": "ff0000",
        },
        Object {
          "colour": "blue",
          "hex": "0000ff",
        },
      ],
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
  ],
  "success": true,
}
`;

exports[`GET all shoes with brand of nike 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "red",
          "hex": "ff0000",
        },
        Object {
          "colour": "blue",
          "hex": "0000ff",
        },
      ],
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
  ],
  "success": true,
}
`;

exports[`GET all shoes with brand of nike and style of basketball 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "red",
          "hex": "ff0000",
        },
        Object {
          "colour": "blue",
          "hex": "0000ff",
        },
      ],
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
  ],
  "success": true,
}
`;

exports[`GET all shoes with collection of hypes 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "red",
          "hex": "ff0000",
        },
        Object {
          "colour": "blue",
          "hex": "0000ff",
        },
      ],
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
  ],
  "success": true,
}
`;

exports[`GET all shoes with non existent brand 1`] = `
Object {
  "data": Array [],
  "empty": true,
  "success": true,
}
`;

exports[`GET all shoes with section of mens 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "red",
          "hex": "ff0000",
        },
        Object {
          "colour": "blue",
          "hex": "0000ff",
        },
      ],
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 2,
      "Section": "mens",
      "Stars": 2,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "green",
          "hex": "00ff00",
        },
      ],
      "description": "dive",
      "name": "jumpers",
      "price": 80,
      "releaseDate": 2098475298,
    },
  ],
  "success": true,
}
`;

exports[`GET all shoes with section of mens and collection of hypes 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "red",
          "hex": "ff0000",
        },
        Object {
          "colour": "blue",
          "hex": "0000ff",
        },
      ],
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
  ],
  "success": true,
}
`;

exports[`GET all shoes with stars over 3 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "red",
          "hex": "ff0000",
        },
        Object {
          "colour": "blue",
          "hex": "0000ff",
        },
      ],
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
  ],
  "success": true,
}
`;

exports[`GET all shoes with style of basketball 1`] = `
Object {
  "data": Array [
    Object {
      "Brand": "nike",
      "Collection": "hypes",
      "ID": 1,
      "Section": "mens",
      "Stars": 3.5,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "red",
          "hex": "ff0000",
        },
        Object {
          "colour": "blue",
          "hex": "0000ff",
        },
      ],
      "description": "run",
      "name": "airss",
      "price": 40,
      "releaseDate": 2938457,
    },
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 2,
      "Section": "mens",
      "Stars": 2,
      "Style": "basketball",
      "colours": Array [
        Object {
          "colour": "green",
          "hex": "00ff00",
        },
      ],
      "description": "dive",
      "name": "jumpers",
      "price": 80,
      "releaseDate": 2098475298,
    },
  ],
  "success": true,
}
`;

exports[`GET single shoe get shoe with id 1 1`] = `
Object {
  "data": Object {
    "Brand": "nike",
    "Collection": "hypes",
    "ID": 1,
    "Section": "mens",
    "Stars": 3.5,
    "Style": "basketball",
    "description": "run",
    "name": "airss",
    "price": 40,
    "releaseDate": 2938457,
  },
  "success": true,
}
`;

exports[`GET single shoe get shoe with id 2 1`] = `
Object {
  "data": Object {
    "Brand": "adias",
    "Collection": "air max",
    "ID": 2,
    "Section": "mens",
    "Stars": 2,
    "Style": "basketball",
    "description": "dive",
    "name": "jumpers",
    "price": 80,
    "releaseDate": 2098475298,
  },
  "success": true,
}
`;

exports[`GET single shoe get shoe with id 3 1`] = `
Object {
  "data": Object {
    "Brand": "adias",
    "Collection": "air max",
    "ID": 3,
    "Section": "womens",
    "Stars": 0,
    "Style": "runners",
    "description": "sprint",
    "name": "kicks",
    "price": 120,
    "releaseDate": 2098475234,
  },
  "success": true,
}
`;

exports[`GET single shoe get shoe with id 4 (doesn't exist) 1`] = `
Object {
  "empty": true,
  "success": true,
}
`;

exports[`PATCH (update) a shoe 1 item updated 1`] = `
Object {
  "data": Object {
    "ID": "1",
    "name": "fly kicks",
  },
  "success": true,
}
`;

exports[`PATCH (update) a shoe 1 item updated 2`] = `
Object {
  "data": Object {
    "Brand": "nike",
    "Collection": "hypes",
    "ID": 1,
    "Section": "mens",
    "Stars": 3.5,
    "Style": "basketball",
    "description": "run",
    "name": "fly kicks",
    "price": 40,
    "releaseDate": 2938457,
  },
  "success": true,
}
`;

exports[`PATCH (update) a shoe 3 items updated 1`] = `
Object {
  "data": Object {
    "ID": "3",
    "description": "runs fast",
    "name": "test shoe",
    "price": 50,
  },
  "success": true,
}
`;

exports[`PATCH (update) a shoe 3 items updated 2`] = `
Object {
  "data": Object {
    "Brand": "adias",
    "Collection": "air max",
    "ID": 3,
    "Section": "womens",
    "Stars": 0,
    "Style": "runners",
    "description": "runs fast",
    "name": "test shoe",
    "price": 50,
    "releaseDate": 2098475234,
  },
  "success": true,
}
`;

exports[`PATCH (update) a shoe no data 1`] = `
Object {
  "error": true,
  "message": "no data sent",
}
`;

exports[`Post add shoe no cover image 1`] = `
Object {
  "data": Object {
    "BrandID": 2,
    "CollectionID": 1,
    "SectionID": 1,
    "StyleID": 1,
    "description": "runs fast",
    "name": "test shoe",
    "price": 50,
    "releaseDate": 29456887,
  },
  "success": true,
}
`;

exports[`Post add shoe no cover image 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 4,
      "Section": "mens",
      "Stars": 0,
      "Style": "runners",
      "colours": Array [],
      "description": "runs fast",
      "name": "test shoe",
      "price": 50,
      "releaseDate": 29456887,
    },
  ],
  "success": true,
}
`;

exports[`Post add shoe no data 1`] = `
Object {
  "error": true,
  "message": "no data sent",
}
`;

exports[`Post add shoe perfect amount of data 1`] = `
Object {
  "data": Object {
    "BrandID": 2,
    "CollectionID": 1,
    "CoverImage": 1,
    "SectionID": 1,
    "StyleID": 1,
    "description": "runs fast",
    "name": "test shoe",
    "price": 50,
    "releaseDate": 29456887,
  },
  "success": true,
}
`;

exports[`Post add shoe perfect amount of data 2`] = `
Object {
  "data": Array [
    Object {
      "Brand": "adias",
      "Collection": "air max",
      "ID": 4,
      "Section": "mens",
      "Stars": 0,
      "Style": "runners",
      "colours": Array [],
      "description": "runs fast",
      "name": "test shoe",
      "price": 50,
      "releaseDate": 29456887,
    },
  ],
  "success": true,
}
`;
